### PR TITLE
add qwen2-vl support

### PR DESCRIPTION
**Pull request by Agoniii**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add Qwen2-VL support

Need to be used with the [mrope support in Mcore](https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/2718).

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- Generate

```shell
python scripts/vlm/qwen2vl_generate.py --load_from_hf
```
- Finetune
mock data:
```shell
torchrun --nproc_per_node 2 scripts/vlm/qwen2vl_finetune.py --gbs=1 --mbs=1 --max_steps=10 --tp_size 2 --devices 2 --log_dir ./outputs
```
real data
```shell
torchrun --nproc_per_node=4 scripts/vlm/qwen2vl_finetune.py \
     --num_nodes=1 --devices=4  --tp_size=4  \
     --gbs=1 --mbs=1  \
     --max_steps=1000 \
     --restore_path=$RESTORE_PATH \
     --projector_type="mcore_mlp" \
     --data_type="qwen2vl" \
     --data_path=$DATA_PATH \
     --image_folder=$IMAGE_FOLDER \
     --log_dir="/workspace/logs"
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: scripts/vlm/qwen2vl_convert_ckpt.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from pathlib import Path
```
Comment (yaoyu-33) [inline scripts/vlm/qwen2vl_convert_ckpt.py:1]: it doesnt need to be a stand alone scirpt. You can put this in doc and also you can put this in the comments of finetune script. Try to keep the folder clear.

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 2]
```diff
@@ -0,0 +1,1096 @@
...
+from nemo.collections.vlm.layer_specs import get_layer_spec_te
+from nemo.collections.vlm.qwen2vl.data.multimodal_tokens import IGNORE_INDEX, IMAGE_TOKEN_INDEX, VIDEO_TOKEN_INDEX
+from nemo.collections.vlm.qwen2vl.model.vision import Qwen2VisionModel
+from nemo.lightning import io
+from nemo.lightning.io.pl import ckpt_to_weights_subdir
+from nemo.lightning.megatron_parallel import MaskedTokenLossReductionWithLossMask
+from nemo.lightning.pytorch.optim import MegatronOptimizerModule, OptimizerModule
+from nemo.utils import logging
+
+MODEL_CONFIG_ATTR = [
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:51]: change to import from neva plz

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 3]
```diff
@@ -0,0 +1,1096 @@
...
+            spatial_patch_size=self.spatial_patch_size,
+            img_h=self.img_h,
+            img_w=self.img_w,
+        )
+
+        return model
+
+
+@dataclass
+class Qwen2VLProjectorConfig(TransformerConfig, io.IOMixin):
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:201]: why this cannot re-use our current `vlm.MultimodalProjectorConfig`?

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 4]
```diff
@@ -0,0 +1,1096 @@
...
+                tokenizer=tokenizer,
+                pre_process=pre_process,
+                post_process=post_process,
+            )
+
+            self.share_embeddings_and_output_weights = self.language_model.share_embeddings_and_output_weights
+            self._language_max_sequence_length = self.language_model.max_sequence_length
+            self._language_is_pipeline_parallel = language_transformer_config.pipeline_model_parallel_size > 1
+            if config.language_model_from_pretrained is not None:
+                sharded_state_dict = dict(state_dict=self.language_model.sharded_state_dict(prefix="module."))
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:373]: can we switch to `neva/base.py`'s `restore_model_weights` func?

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+            freeze_language_model=config.freeze_language_model,
+            freeze_vision_model=config.freeze_vision_model,
+            freeze_vision_projection=config.freeze_vision_projection,
+        )
+
+        self.model_type = ModelType.encoder_or_decoder
+        # This attribute is needed to check if an all-reduce is required
+        # on the word embeddings inside `finalize_model_grads._allreduce_word_embedding_grads`.
+
+        self.vision_model_from_hf = str(self.vision_model.__class__.__module__).startswith("transformers.")
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:405]: do you want to support different HF vit as vision encoder?
If this is not really get used, just remove it.

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+            runtime_gather_output=runtime_gather_output,
+        )  # output shape: [batch_size, seq length, vocab_size]
+
+        if labels is None or loss_mask is None:
+            return output
+        else:
+            return output, final_loss_mask.contiguous()
+
+    # override _preprocess_data() in megatron-lm/megatron/core/models/multimodal/llava_model.py
+    def _preprocess_data(
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:762]: we recently have updated this func in neva a bit. Can you double check to see if anything need to be updated.

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 7]
```diff
@@ -0,0 +1,1096 @@
...
+        if attention_mask is None:
+            final_attention_mask = attention_mask
+        else:
+            # TODO: use no cache opiton for now
+            past_seen_tokens = 0  # past_key_values.get_seq_length() if past_key_values is not None else 0
+            cache_position = torch.arange(
+                past_seen_tokens, past_seen_tokens + combined_embeddings.shape[0], device=combined_embeddings.device
+            )
+
+            final_attention_mask = self._update_causal_mask(
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:740]: are we using a special mask here? i.e. not triangular causal mask?
this mask wont do anything in TE right now.
Our mask type is default to causal, so it will always be causal and ignore the mask inputed here.

File: nemo/collections/vlm/qwen2vl/model/vision.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            cu_seqlens_kv=cu_seqlens,
+            max_seqlen_q=max_seqlen,
+            max_seqlen_kv=max_seqlen,
+            qkv_format='thd',
+        )
+
+    def forward(
+        self, x: torch.Tensor, grid_thw: torch.Tensor, attention_mask: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        """Forward function of the CLIP ViT Model. This function passes the input tensors
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/vision.py:163]: double check the docstrings? seems not matching with the code

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 3 above)
Comment (Agoniii) [reply nemo/collections/vlm/qwen2vl/model/base.py:201]: When using `mlp2x_gelu`, the weight of second Linear  is [ffn_hidden_size, hidden_size]. 
But in MultimodalProjectorConfig, the weight of second Linear  is [hidden_size, hidden_size].

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 7 above)
Comment (chrjxj) [reply nemo/collections/vlm/qwen2vl/model/base.py:740]: correct. it is still triangular causal mask. we could keep `final_attention_mask` none when passing to self.language_model()

File: nemo/collections/vlm/qwen2vl/data/config.p [hunk 9]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/vlm/qwen2vl/data/config.p:1]: actually the config.py exists, so this need to be removed.

File: nemo/collections/vlm/qwen2vl/data/conversation.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+
+default_conversation = conv_qwen2vl
+conv_templates = {
+    "default": conv_qwen2vl,
+    "qwen2vl": conv_qwen2vl,
+    "chatml_direct": conv_chatml_direct,
+}
+
+if __name__ == "__main__":
+    print(default_conversation.get_prompt())
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/conversation.py:290]: Can you please search the page. For `print` and `logger` plz change to 
`from nemo.utils import logging` and `logging.info`

for not used print, please remove

File: nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Callable, Optional
+
+
+@dataclass
+class MultiModalToken:
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py:20]: Are these tokens classes getting used? Doesnt need to fit neva design here.
if not used you can skip the data classes. Only keep
 "Constants for token indexing and special token mapping"

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 3 above)
Comment (yaoyu-33) [reply nemo/collections/vlm/qwen2vl/model/base.py:168]: ah okay @Agoniii  I suggest you modify `vlm.MultimodalProjectorConfig` to match your design here. I think it makes sense to update middle layers to use ffn hidden instead of hidden all the way.

Then you can use vlm.MultimodalProjectorConfig directly

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 12]
```diff
@@ -0,0 +1,1040 @@
...
+
+            language_embeddings = self.language_model.embedding(
+                input_ids=input_ids_text, position_ids=None
+            )  # [decoder_seq_len, b, h_language]
+
+            if self.sequence_parallel_lm:
+                # Gather the language embeddings back.
+                # We use the full embedding to insert image embeddings
+                # and then scatter to avoid load imbalance.
+                language_embeddings = gather_from_sequence_parallel_region(
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:660]: this is no longer needed. Just add https://github.com/NVIDIA/NeMo/blob/c22b3da71e108f7d1b50cb79abcacdfaa4c48aa6/nemo/collections/vlm/neva/model/base.py#L282

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 13]
```diff
@@ -0,0 +1,1040 @@
...
+            inference_params=inference_params,
+            runtime_gather_output=runtime_gather_output,
+        )  # output shape: [batch_size, seq length, vocab_size]
+
+        if labels is None or loss_mask is None:
+            return output
+        else:
+            return output, final_loss_mask.contiguous()
+
+    # override _preprocess_data() in megatron-lm/megatron/core/models/multimodal/llava_model.py
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:696]: Can you add a bit more comments what has been done in this function? I think Luke has a nice comparison table in our thread. just update a bit and comment it here would be helpful!

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 14]
```diff
@@ -0,0 +1,1040 @@
...
+            self.vision_model.set_input_tensor(input_tensor[0])
+        elif self.pre_process:
+            self.encoder_hidden_state = input_tensor[0]
+        else:
+            self.language_model.set_input_tensor(input_tensor[0])
+
+    # pylint: disable=C0301
+    # Copied from transformers.models.phi3.modeling_phi3.Phi3Model._update_causal_mask
+    # copied and adapted from https://github.com/huggingface/transformers/blob/5523e38b553ff6c46b04d2376870fcd842feeecc/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L1189
+    def _update_causal_mask(
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:846]: is this still needed?

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 15]
```diff
@@ -0,0 +1,1040 @@
...
+        inference_params: Optional[InferenceParams] = None,
+        pixel_values: Optional[torch.Tensor] = None,
+        pixel_values_videos: Optional[torch.FloatTensor] = None,
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        video_grid_thw: Optional[torch.LongTensor] = None,
+        image_token_index: Optional[int] = IMAGE_TOKEN_INDEX,
+        video_token_index: Optional[int] = VIDEO_TOKEN_INDEX,
+        runtime_gather_output: Optional[bool] = None,
+    ) -> torch.Tensor:
+        """Forward function of the LLaVA model.
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:543]: plz update this docstring to match the args.

File: nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py (hunk 11 above)
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py:20]: these special tokens comes from https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct/blob/main/tokenizer_config.json, we could remove dataclass definitions but directly use value and string in mapping.

File: nemo/collections/vlm/qwen2vl/data/config.p (hunk 9 above)
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/config.p]: Originally, we want two DataConfig, one for image, the other for video, almost the same as neva. but finally, we combine it into one.
But Qwen2VLDataConfig() is used by qwen2vl_finetune.py. if we remove it, then where to put it? @yaoyu-33

File: nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py (hunk 11 above)
Comment (maureen1111) [reply nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py:20]: Have removed the classes and code will updated later.

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 3 above)
Comment (Agoniii) [reply nemo/collections/vlm/qwen2vl/model/base.py]: I have updated, pls check

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 13 above)
Comment (Agoniii) [reply nemo/collections/vlm/qwen2vl/model/base.py:632]: updated, pls check

File: nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py (hunk 11 above)
Comment (Agoniii) [reply nemo/collections/vlm/qwen2vl/data/multimodal_tokens.py:20]: updated, pls check

**Review by yaoyu-33 — APPROVED** [review]

**Review by yaoyu-33 — DISMISSED** [review]

**Review by yaoyu-33 — DISMISSED** [review]

**Review by yaoyu-33 — APPROVED** [review]
